# 45. 프로미스

콜백 함수의 처리보다 더 나은 비동기 처리 패턴으로 `프로미스(Promise)`가 도입되었다.

단점은 보완하고 비동기 처리 시점을 명확하게 표현할 수 있다

## 45.1 비동기 처리를 위한 콜백 패턴의 단점

### 45.1.1 콜백 헬

- 비동기 함수 처리를 하다보면 이벤트 발생 시 `태스크 큐`에 저장되어 대기하다가 `콜 스택`이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행 과정을 거친다.

그래서 비동기 함수에서 결과값을 console에 찍었을 때 undefined로 나오는 이유가 위와 같다.

즉, 비동기 함수에 비동기 처리 결과에 대한 후속 처리를 수행하는 콜백 함수를 전달하여 처리한다.

다만, 이러한 과정이 중첩되는 상황들이 있으면 계속 처리해야된다는 단점이 생겨서 복잡도가 증가해지게 된다. 이러한 현상을 `콜백 헬`이라고 한다.

```javascript
get("/step1", (a) => {
  get(`/step2/${a}`, (b) => {
    get(`/step3/${b}`, (c) => {
      get(`/step4/${c}`, (d) => {
        console.log(d);
      });
    });
  });
});
```

### 45.1.2 에러 처리의 한계

그래서 위의 경우의 예시도 있지만, `에러 처리`가 어렵다

```javascript
try {
  setTimeout(() => {
    throw new Error("Error!");
  }, 1000);
} catch (e) {
  // 에러를 캐치하지 못한다
  console.error("캐치한 에러", e);
}
```

에러는 `호출자` 방향으로 전파된다.

위의 경우 setTimeout 함수의 콜백 함수를 호출한 것이 `setTimeout`함수가 아니다.

그렇기에 에러 catch가 안된다.

위와 같은 문제점들로 인해 `Promise`가 등장하였다

## 45.2 프로미스의 생성

Promise 생성자 함수를 `new 연산자`와 함께 Promise 객체를 생성해서 사용한다.

Promise는 인수를 전달받는데, 이 콜백 함수는 `resolve`, `reject` 함수를 받는다

```javascript
// 프로미스 생성
const promise = new Promise((resolve, reject) => {
  // Promise 함수의 콜백 함수 내부에서 비동기 처리를 수행한다.
  if (/* 비동기 처리 성공 */) {
    resolve('result');
  } else { /* 비동기 처리 실패 */
    reject('failure reason');
  }
});
```

| 프로미스 상태 정보 | 의미                                  | 상태 변경 조건                   |
| ------------------ | ------------------------------------- | -------------------------------- |
| pending            | 비동기 처리가 아직 수행되지 않은 상태 | 프로미스가 생성된 직후 기본 상태 |
| fulfilled          | 비동기 처리가 수행된 상태(성공)       | resolve 함수 호출                |
| rejected           | 비동기 처리가 수행된 상태(실패)       | reject 함수 호출                 |

![image](https://github.com/user-attachments/assets/303154ef-de31-4e90-a8d2-8b243a6f467d)


> 프로미스는 비동기 처리 상태와 처리 결과를 관리하는 객체이다

## 45.3 프로미스의 후속 처리 메서드

프로미스 처리 결과를 바탕으로 후속 처리 메서드로 then, catch, finally를 제공한다

즉, 프로미스의 비동기 처리 상태가 `변화하면` 후속 처리 메서드에 인수로 전달한 콜백 함수가 선택적으로 호출된다.

### 45.3.1 Promise.prototype.then

- 첫 번째 콜백 함수는 프로미스가 `fulfilled` 상태(resolve 함수가 호출된 상태)가 되면 호출 된다.
- 두 번째 콜백 함수는 프로미스가 `rejected` 상태(reject 함수가 호출된 상태)가 되면 호출된다.

서로 각각 인자로 순서대로 받는다

```javascript
// fulfilled
new Promise((resolve) => resolve("fulfilled")).then(
  (v) => console.log(v),
  (e) => console.error(e)
); // fulfilled

// rejected
new Promise((_, reject) => reject(new Error("rejected"))).then(
  (v) => console.log(v),
  (e) => console.error(e)
); // Error: rejected
```

### 45.3.2 Promise.prototype.catch

catch 메서드는 프로미스가 `rejected`상태인 경우만 호출된다.

```javascript
// rejected
new Promise((_, reject) => reject(new Error("rejected"))).catch((e) =>
  console.log(e)
); // Error: rejected
```

### 45.3.3 Promise.prototype.finally

fianlly 메서든느 프로미스의 성공 또는 실패와 `상관 없이` 무조건 한 번 호출 된다.

공통적으로 수행해야 할 처리 내용이 있을 때 유용하다

```javascript
new Promise(() => {}).finally(() => console.log("finally")); // finally
```

## 45.4 프로미스의 에러 처리

```javascript
promiseGet("https://jsonplaceholder.typicode.com/todos/1")
  .then((res) => console.xxx(res))
  .catch((err) => console.error(err)); // TypeError: console.xxx is not a function
```

두 번째 콜백 함수 등도 사용할 수 있지만 가독성 측면과 간편하기에 위와 같은 형식으로 에러 처리를 진행하는 것이 좋다.

## 45.5 프로미스 체이닝

콜백 헬을 then, catch, finally 후속 처리로 해결한다.

후속 처리는 언제나 `Promise`객체를 반환하기 때문에 연속적으로 체이닝을 걸어 호출할 수 있다.

다만, 이렇게 체이닝이 많아질 경우 콜백 패턴을 사용하지 않는 것이 아니기 때문에 콜백 헬처럼 모습이 보인다.

그래서 `async, await` 문법을 사용하기도 한다.

## 45.6 프로미스의 정적 메서드

- Promise는 주로 생성자 함수로 사용되지만, 함수도 객체이기에 메서드를 가진다.

### 45.6.1 Promise.resolve / Promise.reject

```javascript
// 배열을 resolve하는 프로미스를 생성
const resolvedPromise = Promise.resolve([1, 2, 3]);
resolved;
```

```javascript
// 에러 객체를 reject하는 프로미스를 생성
const rejectedPromise = Promise.reject(new Error("Error!"));
rejectedPromise.catch(console.log); // Error: Error!
```

### 45.6.2 Promise.all

- 여러 개의 비동기 처리를 모두 `병렬`처리할 때 사용한다.

```javascript
const requestData1 = () =>
  new Promise((resolve) => setTimeout(() => resolve(1), 3000));
const requestData2 = () =>
  new Promise((resolve) => setTimeout(() => resolve(2), 2000));
const requestData3 = () =>
  new Promise((resolve) => setTimeout(() => resolve(3), 1000));

// 세 개의 비동기 처리를 순차적으로 처리
const res = [];
requestData1()
  .then((data) => {
    res.push(data);
    return requestData2();
  })
  .then((data) => {
    res.push(data);
    return requestData3();
  })
  .then((data) => {
    res.push(data);
    console.log(res); // [1, 2, 3] ⇒ 약 6초 소요
  })
  .catch(console.error);
```

- 위의 예시의 경우 `순차적`으로 체이닝이 걸리게 되어 오래 걸리지만,

아래와 같이 Promise.all을 사용했을 때 병렬처리를 진행해서 시간을 단축할 수 있다.

```javascript
const requestData1 = () =>
  new Promise((resolve) => setTimeout(() => resolve(1), 3000));
const requestData2 = () =>
  new Promise((resolve) => setTimeout(() => resolve(2), 2000));
const requestData3 = () =>
  new Promise((resolve) => setTimeout(() => resolve(3), 1000));

Promise.all([requestData1(), requestData2(), requestData3()])
  .then(console.log) // [ 1, 2, 3 ] ⇒ 약 3초 소요
  .catch(console.error);
```

1. resolve 시 처리 순서가 `보장`되며 resolve처리 결과를 차례로 배열에 저장해 Promise를 반환한다.

2. 프로미스가 하나라도 `rejected`상태가 되면 나머지 프로미스가 fulfilled 상태가 되는 것을 기다리지 않고 `즉시 종료`한다.

### 45.6.3 Promise.race

- Promise.all과 비슷하나 모든 프로미스가 fulfilled 상태가 되는 것을 기다리는 것이 아니라 `가장 먼저 fulfilled` 상태가 된 프로미스의 처리 결과를 resolve하는 새로운 프로미스를 반환한다.

```javascript
Promise.race([
  new Promise((resolve) => setTimeout(() => resolve(1), 3000)), // 1
  new Promise((resolve) => setTimeout(() => resolve(2), 2000)), // 2
  new Promise((resolve) => setTimeout(() => resolve(3), 1000)), // 3
])
  .then(console.log) // 3
  .catch(console.log);
```

- reject 반환은 Promise.all과 같다.

### 45.6.4 Promise.allSettled

- 전달받은 프로미스가 모두 settled된 상태 즉, 비동기 처리가 수행된 상태를 말한다.

- resolve, reject 상태와는 상관없이 `모든 프로미스들의 처리 결과`가 담겨있다.

```javascript
Promise.allSettled([
  new Promise((resolve) => setTimeout(() => resolve(1), 2000)),
  new Promise((_, reject) =>
    setTimeout(() => reject(new Error("Error!")), 1000)
  ),
]).then(console.log);
/*
[
  {status: "fulfilled", value: 1},
  {status: "rejected", reason: Error: Error! at <anonymous>:3:54}
]
*/
```

```javascript
[
  // 프로미스가 fulfilled 상태인 경우
  {status: "fulfilled", value: 1},
  // 프로미스가 rejected 상태인 경우
  {status: "rejected", reason: Error: Error! at <anonymous>:3:60}
]
```

## 45.7 마이크로태스크 큐

```javascript
setTimeout(() => console.log(1), 0);

Promise.resolve()
  .then(() => console.log(2))
  .then(() => console.log(3));
```

- 프로미스 후속 처리 메서드도 비동기로 동작하기에 2 -> 3 -> 1순으로 출력된다.

그 이유는 프로미스 후속 처리 메서드의 콜백 함수는 태스크 큐가 아니라 `마이크로태스크 큐`에 저장된다.

우선순위를 따지면 마이크로태스크 큐 > 태스크 큐 이기에 `이벤트 루프`는 콜 스택이 비었을 때 마이크로태스크 큐에서 대기하고 있는 함수를 `먼저` 꺼내서 실행한다

## 45.8 fetch

XMLHttpRequest 객체와 마찬가지로 HTTP 요청 전송 기능을 제공하는 Web API이다.

```javascript
fetch("https://jsonplaceholder.typicode.com/todos/1").then((response) =>
  console.log(response)
);
```

> fetch 함수는 HTTP 응답을 나타내는 Response 객체를 래핑한 Promise 객체를 반환한다.

![image](https://github.com/user-attachments/assets/d31e9150-159f-4415-8855-1a01db0f530c)


fetch 함수가 반환하는 프로미스는 404, 500 같은 HTTP에러가 발생해도 에러를 reject 하지 않고, Reponse객체를 resolve한다.

오프라인 등의 `네트워크 장애`나 `CORS`에러 등의 요청이 완료되지 못한 경우에만 프로미스를 `reject`한다

```javascript
const wrongUrl = "https://jsonplaceholder.typicode.com/XXX/1";

// 부적절한 URL이 지정되었기 때문에 404 Not Found 에러가 발생한다.
fetch(wrongUrl)
  // response는 HTTP 응답을 나타내는 Response 객체다.
  .then((response) => {
    if (!response.ok) throw new Error(response.statusText);
    return response.json();
  })
  .then((todo) => console.log(todo))
  .catch((err) => console.error(err));
```
